packages/core/src/components/overlay2/overlay2.tsx

```diff
@@ -325,14 +306,41 @@ export const Overlay2 = React.forwardRef<OverlayInstance, Overlay2Props>((props,
         }
     }, [isOpen, overlayWillOpen, overlayWillClose, prevIsOpen]);
 
-    // run once on unmount
+    // Important: clean up old document-level event listeners if their memoized values change (this is rare, but
+    // may happen, for example, if a user forgets to use `React.useCallback` in the `props.onClose` value).
+    // Otherwise, we will lose the reference to those values and create a memory leak since we won't be able
+    // to successfully detach them inside overlayWillClose.
     React.useEffect(() => {
+        if (!isOpen || !(canOutsideClickClose && !hasBackdrop)) {
+            return;
+        }
+
+        document.addEventListener("mousedown", handleDocumentMousedown);
+
         return () => {
-            // we need to run cleanup code to remove some event handlers from the overlay element
-            overlayWillClose();
```

Comment:
I wasn't able to find a test case that can assert this because we only call `onClose` from user actions, but I believe this is a function that has a stale closure around old versions of the document event listeners that may have already been removed.

Now we rely on the `useEffect` that adds the event listeners to clean up themselves - but continue calling this for the other overlay cleanup code. I [didn't see any warnings](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener) about calling `removeEventListener` twice/for an already removed handler, and also didn't see any issues with this from my testing, but please flag if this seems bad.

Reply:
> I believe this is a function that has a stale closure around old versions of the document event listeners that may have already been removed.

Agree, overlayWillClose was not a dependency of this effect but itself had dependencies.

> I didn't see any warnings about calling `removeEventListener` twice

Quoting the page you linked:

> Calling `removeEventListener()` with arguments that do not identify any currently registered event listener on the `EventTarget` has no effect.

This should be fine.